### PR TITLE
initial implementation of mouse controls in menu

### LINKED ISSUE #861: Mouse Support for Menus
Would it be out of the scope to have the ability to navigate the menus with the mouse? I think it would be a nice addition to have that makes the port feel that much more fluid.

### PR DESCRIPTION
**Pull request by rfomin**
Thermo bars and multi-select items do not work yet. Not sure how to implement them.

Fixes #861

### CODE REVIEW

**Comment by rfomin** [comment]
https://user-images.githubusercontent.com/5077629/227307653-f79f88a9-7445-4031-b9c8-7fff768d45ae.mp4

**Comment by fabiangreffrath** [comment]
Very cool, thank you so much!

Regarding thermo bars and multi items, how about left click increases their value by one with a wrap-around when the max value is reached? Though, that would be inconsistent with current keyboard behavior.

File: src/i_video.c [hunk 1]
```diff
@@ -304,11 +304,6 @@ static boolean MouseShouldBeGrabbed(void)
    if (!window_focused)
       return false;
    
-   // always grab the mouse when full screen (dont want to 
-   // see the mouse pointer)
-   if (fullscreen)
```
Comment (fabiangreffrath) [inline src/i_video.c:309]: `if (fullscreen && !menuactive)` maybe?

I wouldn't want to see the mouse e.g. on the intermission screen.

**Comment by fabiangreffrath** [comment]
Blinking arrows next to "<- prev" and "next ->" should probably not get printed.

**Comment by fabiangreffrath** [comment]
> ~Regarding thermo bars and multi items, how about left click increases their value by one with a wrap-around when the max value is reached? Though, that would be inconsistent with current keyboard behavior.~

Forget that. How about, clicking thermo bars and multi items puts them into "change" mode, and then moving the mouse horizontally changes their value (left = decrease, right = increase)? Also, it is currently impossible to leave "change" mode by right mouse click.

**Comment by rfomin** [comment]
> Forget that. How about, clicking thermo bars and multi items puts them into "change" mode, and then moving the mouse horizontally changes their value (left = decrease, right = increase)? Also, it is currently impossible to leave "change" mode by right mouse click.

I'm thinking, maybe we don't need a "change" mode (`setup_select` in the code) with the mouse at all? Also, we don't need `setup_select` with the keboard/gamepad for `S_YESNO` items, but it may have been done for consistency.

I tried to get rid of `setup_select` but couldn't. Maybe it's time to refactor that giant `M_Responder()` function, or is it not worth it?

File: src/i_video.c (hunk 1 above)
Comment (rfomin) [reply src/i_video.c:309]: How about this fix? https://github.com/fabiangreffrath/woof/pull/956/commits/4552a9982bad8e73ee9d86f5eb0177dc7bc9e6a1

**Comment by rfomin** [comment]
> Blinking arrows next to "<- prev" and "next ->" should probably not get printed.

Fixed in https://github.com/fabiangreffrath/woof/pull/956/commits/51be9ae8859334dab60bfa2bbf7700f0cb9601e9

**Comment by fabiangreffrath** [comment]
> I'm thinking, maybe we don't need a "change" mode (`setup_select` in the code) with the mouse at all (or at least for thermo and `S_YESNO` items)? Also, we don't need `setup_select` with the keboard/gamepad for `S_YESNO` items, but it may have been done for consistency.

I think we need to preserve the "change" mode, it is anchored too deeply in my muscle memory. For example, I don't have it in Crispy's Crispness menu and I am always puzzled why the items changed even though I selected them "only once".

> I tried to get rid of `setup_select` but couldn't. Maybe it's time to refactor that giant `M_Responder()` function, or is it not worth it?

If we have a chance to rewrite the entire menu code in a semantic way with self-arranging pages, I'd say let's throw out all the old cruft. But we are not there (yet).

Sure, `M_Responder()` is a beast. Would it help if we could factor out some of the sub-menu specific code maybe?

**Comment by rfomin** [comment]
> I think we need to preserve the "change" mode, it is anchored too deeply in my muscle memory.

I suggest keeping "change" mode for keyboard/gamepad, but disable it for mouse controls. There is no such muscle memory yet 😄 Also, I still have not figured out how to make multiselect and thermo items in "change" mode with mouse controls. By the way, if you have ideas, please feel free to commit to this PR.

> Sure, `M_Responder()` is a beast. Would it help if we could factor out some of the sub-menu specific code maybe?

We can factor out all `setup_active` code.

**Comment by fabiangreffrath** [comment]
So, how do we proceed here?

**Comment by fabiangreffrath** [comment]
> I don't know. Seems like too many changes for a minor feature, if we want to do it properly.

Hm, I don't even think it's so many changes. Actually, I find it quite nice. The only question that's left IIRC is how to proceed with multi-choice items and thermo bars, right?

**Comment by rfomin** [comment]
> The only question that's left IIRC is how to proceed with multi-choice items and thermo bars, right?

Yes, and I think to make them properly, we need a lot of changes. For example, turn off the “change” mode, add moving the thermobar with the mouse cursor, remake the multi-select item in the form of a drop-down list.

**Comment by fabiangreffrath** [comment]
> > The only question that's left IIRC is how to proceed with multi-choice items and thermo bars, right?
> 
> Yes, and I think to make them properly, we need a lot of changes. For example, turn off the “change” mode, add moving the thermobar with the mouse cursor, remake the multi-select item in the form of a drop-down list.

Another idea: How about we allow to "drag" thermobars and multi-items with the mouse? That would be, you would have to click-hold and then move the mouse to change their value. TO the right would increase, to the left would decrease.

**Comment by rfomin** [comment]
Deprecated by https://github.com/fabiangreffrath/woof/commit/1503cfed67833f76b75186933a81900475c20e23

**Pull request closed without merging**